packages/plugin-hardhat/src/index.ts

```diff
@@ -110,13 +113,45 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE, async (args: RunCompilerArgs, hre, runSup
   const { isFullSolcOutput } = await import('./utils/is-full-solc-output');
   if (isFullSolcOutput(output)) {
     const decodeSrc = solcInputOutputDecoder(args.input, output);
-    const validations = validate(output, decodeSrc, args.solcVersion);
+
+    const namespacedInput = makeNamespacedInput(args.input, output);
+    const { output: namespacedOutput } = await runSuper({ ...args, input: namespacedInput });
+    checkNamespacedCompileErrors(namespacedInput, namespacedOutput);
+
+    const validations = validate(output, decodeSrc, args.solcVersion, args.input, namespacedOutput);
     await writeValidations(hre, validations);
   }
 
   return { output, solcBuild };
 });
 
+/**
+ * Checks for compile errors in the modified contracts for namespaced storage.
+ * If errors are found, throws an error with the compile error messages, and logs
+ * the modified contract source code as debug.
+ */
+function checkNamespacedCompileErrors(namespacedInput: SolcInput, namespacedOutput: SolcOutput) {
+  const errors = [];
+  if (namespacedOutput.errors !== undefined) {
+    for (const error of namespacedOutput.errors) {
+      if (error.severity === 'error') {
+        const msg = error.formattedMessage;
+
+        debug('Compile error in modified contract for namespaced storage.');
+        debug('Error:', msg);
+        if (error.sourceLocation !== undefined && error.sourceLocation.file in namespacedInput.sources) {
+          debug('Modified contract source:', namespacedInput.sources[error.sourceLocation.file].content);
```

Comment:
This will be better I think:

```suggestion
          debug('Modified contract source:', JSON.stringify(namespacedInput.sources[error.sourceLocation.file].content));
```

Reply:
Removed debug messages as mentioned in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/876#discussion_r1333201777